### PR TITLE
Feature/waypointy translations

### PR DESCRIPTION
**Pull request by boly38**
Translations for country and cache types #32 

requirements: 
* [x] check travis++
* (auto) generate message.po using script
* put it on rec before prod :)

NB/
* suggested "fieldset" html element has not been included (not so beautifull in fact..)
* including flag image could rely on a new issue

### CODE REVIEW

File: website/lib/Waypointy.class.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        if (!isset($waypoint) || $waypoint == '') {
+            throw new Exception('Waypoint expected');
+        }
+        $sqlquery = 'SELECT wp.`waypoint`, wp.`lat`, wp.`lon`, wp.`name`, wp.`owner`, wp.`typ`, wpt.`cache_type`, wp.`link`, wp.`kraj`, wpc.`country`'
+               .' FROM `gk-waypointy` wp'
+               .' JOIN `gk-waypointy-country` wpc ON wp.`kraj` = wpc.`kraj`'
+               .' JOIN `gk-waypointy-type` wpt ON wp.`typ` = wpt.`typ`'
+               .' WHERE wp.`waypoint`=\'%s\''
+               .' LIMIT 1';
+        $query = sprintf($sqlquery, mysqli_real_escape_string($this->dblink, $waypoint));
```
Comment (kumy) [inline website/lib/Waypointy.class.php]: It would be even better to use [Sql Prepared Statements](http://php.net/manual/en/mysqli.quickstart.prepared-statements.php).

What about: (adapted from the doc)
```
if (!($stmt = $this->dblink->prepare('SELECT wp.`waypoint`, wp.`lat`, wp.`lon`, wp.`name`, wp.`owner`, wp.`typ`, wpt.`cache_type`, wp.`link`, wp.`kraj`, wpc.`country`'
               .' FROM `gk-waypointy` wp'
               .' JOIN `gk-waypointy-country` wpc ON wp.`kraj` = wpc.`kraj`'
               .' JOIN `gk-waypointy-type` wpt ON wp.`typ` = wpt.`typ`'
               .' WHERE wp.`waypoint`=?'
               .' LIMIT 1'))) {
    throw new Exception('Prepare failed: ('.$this->dblink->errno.') '.$this->dblink->error);
}

if (!$stmt->bind_param("s", $waypoint)) {
    throw new Exception('Binding parameters failed: ('.$stmt->errno.') '.$stmt->error);
}

if (!$stmt->execute()) {
    throw new Exception('Execute failed: ('.$stmt->errno.') '.$stmt->error);
}

if (!$stmt->bind_result($this->waypoint, $this->lat, $this->lon, $this->name, $this->owner, $this->typ, $this->cache_type, $this->cache_link, $this->kraj, $this->country)) {
    throw new Exception('Binding output parameters failed: ('.$stmt->errno.') '.$stmt->error);
}

if ($stmt->fetch()) {
    return true;
}
return false;
```

File: website/lib/Waypointy.class.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    public function getByName($waypointName)
+    {
+        if (!isset($waypointName) || $waypointName == '') {
+            throw new Exception('Waypoint name expected');
+        }
+        $sqlquery = 'SELECT wp.`waypoint`, wp.`lat`, wp.`lon`, wp.`name`, wp.`owner`, wp.`typ`, wpt.`cache_type`, wp.`link`, wp.`kraj`, wpc.`country`'
+            .' FROM `gk-waypointy` wp'
+            .' JOIN `gk-waypointy-country` wpc ON wp.`kraj` = wpc.`kraj`'
+            .' JOIN `gk-waypointy-type` wpt ON wp.`typ` = wpt.`typ`'
+            .' WHERE wp.`name` LIKE \'%%%s%%\''
```
Comment (kumy) [inline website/lib/Waypointy.class.php]: And here, [use for example](https://stackoverflow.com/a/36593020/944936) `CONCAT('%',?,'%')` (with prepared statement too)

File: website/szukaj-ajax.php [hunk 3]
```diff
@@ -8,6 +8,45 @@
...
+ * json result (tresc) to send when ONE cache is selected.
+ */
+function cacheOk($cache_link, $name, $owner, $cache_type, $cache_country, $countryCode)
+{
+    $flagPlaceholder = '';
+    if ($countryCode !== '') {
+        $flagPlaceholder = '<img src="'.CONFIG_CDN_IMAGES.'/country-codes/'.$countryCode.'.png" width="16" height="11" alt="'._('flag').'"/>';
+    }
+    if ($owner !== '') {
+        $ownerPlaceholder = '('._('by').' '.$owner.')';
```
Comment (kumy) [inline website/szukaj-ajax.php]:
```
$ownerPlaceholder = '('.sprintf(_('by %s'), $owner).')';
```

File: website/szukaj-ajax.php [hunk 4]
```diff
@@ -8,6 +8,45 @@
...
+    }
+
+    $result = '<img src="'.CONFIG_CDN_IMAGES.'/icons/ok.png" alt="'._('OK').'" width="16" height="16" /> <a href="'.$cache_link.'" target="_opencaching">'.$name.'</a> '.$ownerPlaceholder.'<br />';
+    $result .= _('cache type').': '._($cache_type).'<br/>';
```
Comment (kumy) [inline website/szukaj-ajax.php]: $result .= sprintf(_('cache type: %s'), _($cache_type)).'<br/>';

File: website/szukaj-ajax.php [hunk 5]
```diff
@@ -74,48 +113,67 @@
...
-             sprintf(_('<a href="%s">Learn more about hiding geokrets in GC caches</a>'), $config['adres'].'help.php#locationdlagc');
-            $return['lat'] = '';
-            $return['lon'] = '';
-            echo json_encode($return);
+            $lat = $waypointy->lat;
+            $lon = $waypointy->lon;
+
+            if (!empty($lat) and !empty($lon)) {
+                $country = $waypointy->country;
+                $countryCode = ''; // @TODO
```
Comment (kumy) [inline website/szukaj-ajax.php]: Really a `TODO`?

File: website/szukaj-ajax.php [hunk 6]
```diff
@@ -74,48 +113,67 @@
...
+            }
 
             // listing
-            $sql = "SELECT `waypoint`, `name`, `owner`, `typ`, `kraj`, `link`, `lat`, `lon` FROM `gk-waypointy` WHERE `name` LIKE '%".mysqli_real_escape_string($link, $_REQUEST['NazwaSkrzynki'])."%' LIMIT 4";
+            $sql = "SELECT `waypoint`, `name`, `owner`, `typ`, `kraj`, `link`, `lat`, `lon` FROM `gk-waypointy` WHERE `name` LIKE '%".mysqli_real_escape_string($link, $_REQUEST['NazwaSkrzynki'])."%' LIMIT $maxCache";
             $result = mysqli_query($link, $sql);
             while ($row = mysqli_fetch_array($result)) {
                 list($waypoint, $name, $owner, $typ, $kraj, $cache_link, $lat, $lon) = $row;
-                $return['tresc'] .= "<span class='bardzomale'><a href='$cache_link'>$name</a> ($owner) - <a href=\"#\" onclick=\"document.getElementById('wpt').value = '$waypoint'; sprawdzWpt(); return false;\">$waypoint</a></span><br />";
+                $return['tresc'] .= "<a href=\"#\" onclick=\"document.getElementById('wpt').value = '$waypoint'; sprawdzWpt(); return false;\">$waypoint</a> - <span class='bardzomale'><a href='$cache_link' target='_opencaching'>$name</a> ("._('by')." $owner)</span><br />";
```
Comment (kumy) [inline website/szukaj-ajax.php]:
```
$return['tresc'] .= "<a href=\"#\" onclick=\"document.getElementById('wpt').value = '$waypoint'; sprawdzWpt(); return false;\">$waypoint</a> - <span class='bardzomale'><a href='$cache_link' target='_opencaching'>$name</a> (".sprintf(_('by %s'), $owner).")</span><br />";
```

**Comment by kumy** [comment]
I've removed the codacy sql check as it checks for PSQL syntax. Other errors seems related to braces move (code reformat). So let's ignore them.

File: website/funkcje.js [hunk 7]
```diff
@@ -144,7 +144,7 @@ function sprawdzNazwe() {
 
   var poleOut = doc.getElementById(poleOutId);
 
-  if (poleInValue.length >= 5) {
+  if (poleInValue.length >= 1) {
```
[the remark this answers is not in the document]
Comment (boly38) [reply website/funkcje.js]: search always limited by ajax. I've done this to get translated error message (from php instead of js)
I agree that it's better to return error asap, will try to change it

File: website/templates/waypointy-translations.html [hunk 8]
```diff
@@ -0,0 +1,98 @@
+<html><head></head><body>
```
[the remark this answers is not in the document]
Comment (boly38) [reply website/templates/waypointy-translations.html]: yrp

File: website/lib/Waypointy.class.php (hunk 1 above)
Comment (boly38) [reply website/lib/Waypointy.class.php]: like it, thanks for this sample. (CR to continue tomorrow)

**Comment by boly38** [comment]
I will not fix codacy check that rely on php_cs update only 
consider them out of this PR scope (like short id for example) as this must be tested to avoid new bug (as we have no automated tests for this part)

**Comment by kumy** [comment]
> I will not fix codacy check that rely on php_cs update only
consider them out of this PR scope 

I agree, this is what I said in previous comment ;)

File: docker/mariadb/gk-waypointy-country.sql [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+('Portugalia',	'Portugal'),
+('Rosja',	'Russia'),
+('Ruanda',	'Rwanda'),
+('RumÃ¤nien',	'Romania'),
+('Rumänien',	'Romania'),
+('Rumunia',	'Romania'),
+('Russische FÃ¶deration',	'Russia'),
+('Russische Föderation',	'Russia'),
+('SÅ‚owacja',	'Slovakia'),
+('SÃ¼dgeorgien und die SÃ¼dlichen Sandwichinseln',	'South Georgia and the South Sandwich Islands'),
```
Comment (kumy) [inline docker/mariadb/gk-waypointy-country.sql]: Hum, there are many encoding problems in this file... In production database, this entry is `Südgeorgien und die Südlichen Sandwichinseln`

File: website/lib/Waypointy.class.php [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        $action = 'Waypointy::getByWaypoint';
+        if (!isset($waypoint) || $waypoint == '') {
+            throw new Exception($action.' waypoint expected');
+        }
+        if (!($stmt = $this->dblink->prepare(self::SELECT_WAYPOINT
+                       .' WHERE wp.`waypoint`=?'
+                       .' LIMIT 1'))) {
+            throw new Exception($action.' prepare failed: ('.$this->dblink->errno.') '.$this->dblink->error);
+        }
+        if (!$stmt->bind_param('s', $waypoint)) {
```
Comment (kumy) [inline website/lib/Waypointy.class.php]: Great!!! :+1:

File: docker/mariadb/database-create-geokrety-db.sql [hunk 11]
```diff
@@ -306,21 +326,21 @@ CREATE TABLE `gk-ostatnieruchy` (
...
-  `logtype` enum('0','1','2','3','4','5','6') COLLATE utf8_polish_ci DEFAULT '0' COMMENT '0=drop, 1=grab, 2=comment, 3=met, 4=arch, 5=dip',
-  `username` varchar(20) COLLATE utf8_polish_ci,
+  `logtype` enum('0','1','2','3','4','5','6') CHARACTER SET utf8 COLLATE utf8_polish_ci DEFAULT '0' COMMENT '0=drop, 1=grab, 2=comment, 3=met, 4=arch, 5=dip',
+  `username` varchar(20) CHARACTER SET utf8 COLLATE utf8_polish_ci,
   `timestamp` timestamp NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
-  `app` varchar(16) COLLATE utf8_polish_ci DEFAULT 'www' COMMENT 'source of the log',
-  `app_ver` varchar(16) COLLATE utf8_polish_ci COMMENT 'apploction version/codename'
-) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_polish_ci;
+  `app` varchar(16) CHARACTER SET utf8 COLLATE utf8_polish_ci DEFAULT 'www' COMMENT 'source of the log',
+  `app_ver` varchar(16) CHARACTER SET utf8 COLLATE utf8_polish_ci COMMENT 'apploction version/codename'
```
Comment (kumy) [inline docker/mariadb/database-create-geokrety-db.sql]: Small typo (existing before) "apploction" "application"

**Comment by kumy** [comment]
Except encoding problems in file `gk-waypointy-country.sql` overall looks good. I will try to deploy on rec (without merging yet)

**Comment by kumy** [comment]
Deployed on rec. Crowdin strings not imported until merged on master.

**Comment by boly38** [comment]
seems not the case
4e737e5 reportrf on footer

**Comment by kumy** [comment]
New tables created and populated also on prod, so we're ready ;)

**Comment by kumy** [comment]
@boly38 can you check if the result is fine on rec, then we'll merge and deploy on prod :)

**Comment by boly38** [comment]
on rec we have latest master commit and not the branch we are taking about.. so I can't test :(

**Comment by kumy** [comment]
You're right, I forget the new upgrade procedure ;) it's ok now

**Comment by boly38** [comment]
seems there is some issue to review:
- search "bon voyage" + button "Verifier"
- on rec we have 3 results: 
- first one (aka. GA1781) has no owner, and "by .." must not be displayed 
- click on second one (aka "OC006E - Bon voyage (by cezanne)") is OK
- but click on first one is not ok: no ajax result.

I've no time to contribute right now, but I must go back on this branch to fix this 2 issues.

**Comment by boly38** [comment]
Note about GA1781:
* seems GA prefix means "coming from geocaching australia"
* geocaching australia are result of [wpt](https://github.com/geokrety/geokrety-scripts/blob/master/waypointy/wpt/wpt) script
* I've no short way to test this script locally but maybe target source format is now slightly different, for example, I was unable to fine `1522.070` sentence in the source of  [gpx](http://geocaching.com.au/caches/gca.gpx)

@kumy do you confirm ? 
BTW I really need this data to be able reproduce the issue :)

**Comment by kumy** [comment]
@boly38 sorry I didn't follow project for some time.
May you confirm your latest issues with GA are now fixed thanks to fix for #79?
If yes, we may merge soon.

**Comment by boly38** [comment]
hello
no I will need your help to reproduce the issue
just a data extract 
insert related to

**Comment by kumy** [comment]
@boly38 didn't you closed the issue by mistake?

**Comment by kumy** [comment]
Sorry for the long delay.
Here is the sql statement you requested:

```
INSERT INTO `gk-waypointy` (`waypoint`, `lat`, `lon`, `alt`, `country`, `name`, `owner`, `typ`, `kraj`, `link`, `status`, `timestamp`) VALUES
('GA1781', -42.83858, 147.28862, -2000, 'au', 'Bon Voyage Swampy!!', '', '', '', 'http://geocaching.com.au/cache/GA1781', 1, '2010-09-02 04:37:07');
```

**Comment by kumy** [comment]
Also, I'm filling rec database with prod database.

**Comment by boly38** [comment]
thx for the insert I will try to reproduce asap

yep I closed by error. Mobile version is too much sensitive 😅

**Comment by boly38** [comment]
just rebase + push fix
- [] to test en rec!

**Comment by kumy** [comment]
@boly38 commit 2e28d79 pushed on rec.

**Comment by boly38** [comment]
I just pushed a fix to handle empty cache type. Sorry to miss that point. Testing context not super easy; so having rec to validate is good point :)

**Comment by kumy** [comment]
Will refresh rec tonight. Btw, I ll try to build a special instance just for you, the idea is to have a branch called boly38 dedicated for you and auto updated

On June 14, 2018 1:13:31 PM GMT+02:00, Boly38 <notifications@github.com> wrote:
>I just pushed a fix to handle empty cache type. Sorry to miss that
>point. Testing context not super easy; so having rec to validate is
>good point :)
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/geokrety/geokrety-website/pull/77#issuecomment-397258773

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by kumy** [comment]
@boly38 rec refreshed

**Comment by boly38** [comment]
thx. I've doubt for use case : got more than 4 results..

**Comment by kumy** [comment]
> I've doubt for use case : got more than 4 results..

Sorry what do you mean?

BTW, I'm proud to tell you that you can test your code using branch boly38 and this url https://boly38.gk.kumy.org/

I'll revert back the rec.geokrety.org to master branch.

**Comment by boly38** [comment]
okkk that's fantastic 🤘

for doubt, nevermind: I need to retest from workstation (instead of mobile) and see code. I think that there must be dedicated message when got x results but only 4 first  are displayed. will see that tomorrow.

**Comment by boly38** [comment]
to go deeper into details (that just for information, wont to enforce related action :p)
would be really usefull to move "prod" and "rec" tags (or branches) for deployed things (like you do for "boly38" one)
- in fact 'master' is not PROD deployed version (and maybe not always 'rec' deployed version too..)
- footer commit SHA1 is not suficent while a given commit could diseapear on force push action

**Comment by boly38** [comment]
@kumy 

your new 'boly38' _dedicated deploy mechanism_ is faster than my local update ! I'm jealous !!
 ;)


----

@kumy @harrieklomp 

ok out of misstake, I think I'm done now

Maybe could you play a little bit with the fresh version onto https://boly38.gk.kumy.org/ (ruchy UI > waypointy search)

Tell me if you encounter some strange behaviours.

PS/ notice that some label are not already on crowdin (and so translated in your language) for example the yellow sentence in my previous comment, or field description / values (cache-type, country). But there is translation placeholder for them.

**Comment by kumy** [comment]
@boly38 What about adding some iconology to make the difference between link permitting to fill the form vs link to the cache details?

screenshot from 2018-06-15 20-14-49

**Comment by boly38** [comment]
yes I like your idea
you could commit on the branch
else I wont work on it until monday

**Comment by kumy** [comment]
New commit pushed on branch boly38 ;)

**Comment by kumy** [comment]
@boly38 really long history in this PR, do you agree to merge on master? I tested the feature and looks good for me.

**Comment by boly38** [comment]
sure !
do not forget geokrety-scripts side
learned a lot
thanks for your help along this PR story 🤘

**Comment by kumy** [comment]
Good work @boly38  :metal:
